src/Service/AbstractRESTService.php

```diff
@@ -0,0 +1,338 @@
+    {
+        if (!array_key_exists($name, $this->options)) {
+            throw new ServiceException(sprintf('No such option: %s', $name));
+        }
+
+        return $this->options[$name];
+    }
+
+    public function isTestMode(): bool
+    {
+        return (bool)$this->options['test_mode'];
+    }
+
+    /**
+     * Call a REST service endpoint.
+     *
+     * @param string $method
+     * @param string $url
+     * @param array $options
+     * @return ResponseInterface
+     * @throws ServiceException
+     */
+    protected function call(string $method, string $url, array $options): ResponseInterface
+    {
+        $samlToken = $this->getSAMLToken();
+        $accessToken = $this->getAccessToken($samlToken);
+
+        // @todo Can we just generate the transaction id?
+        if (!isset($options['transactionId'])) {
+            throw new ServiceException('Missing transaction id');
+        }
+        $transactionId = $options['transactionId'];
+        unset($options['transactionId']);
+
+        $transactionTid = $options['transactionTid'] ?? new DateTimeImmutable();
+        if ($transactionTid instanceof DateTimeInterface) {
+            $transactionTid = Serializer::formatDateTimeZulu($transactionTid);
+        }
+        unset($options['transactionTid']);
+
+        $headers = (isset($options['headers']) && is_array($options['headers'])) ?  $options['headers'] : [];
+        unset($options['headers']);
+
+        return $this->request(
+            $method,
+            $url,
+            [
+                'headers' => [
+                        'Authorization' => sprintf('%s %s', $accessToken['token_type'], $accessToken['access_token']),
+                        'x-TransaktionsId' => $transactionId,
+                        'x-TransaktionsTid' => $transactionTid,
+                    ] + $headers,
+            ]
+            +$options
+        );
+    }
+
+    private function client(): HttpClientInterface
+    {
+        if (null === $this->client) {
+            $this->client = new CurlHttpClient([
+            ]);
+        }
+
+        return $this->client;
+    }
+
+    private function request(string $method, string $url, array $options): ResponseInterface
+    {
+        // Write certicates to temporary files.
+        $certificateFilename = tempnam(sys_get_temp_dir(), 'cert');
+        file_put_contents($certificateFilename, $this->getCertificate());
+
+        $privateKeyFilename = tempnam(sys_get_temp_dir(), 'cert');
+        file_put_contents($privateKeyFilename, $this->getPrivateKey());
+
+        return $this->client()->request($method, $url, $options + [
+                'local_cert' => $certificateFilename,
+                'local_pk' => $privateKeyFilename,
+                'on_progress' => function (int $dlNow, int $dlSize, array $info) use (&$certificateFilename, &$privateKeyFilename): void {
+                    // Delete temporary certificate files when receiving response headers.
```

Comment:
Are temporary files not temporary? Is this necessary?

Reply:
No, temporary files are not temporary and must be deleted manually. I don't understand why this is the case. Maybe I don't understand what “temporary” means …